nixos/modules/services/matrix/synapse/workers.nix

```diff
@@ -0,0 +1,198 @@
+{ commonConfigFile
+, pluginsEnv
+, format
+}:
+
+{ config, lib, pkgs, ... }: let
+  format = pkgs.formats.yaml { };
+
+  cfg = config.services.matrix-synapse;
+
+  inherit (lib) mkOption types literalExpression;
+
+  # checks if given listener configuration has type as a resource
+  isListenerType = type: l: lib.any (r: lib.any (n: n == type) r.names) l.resources;
+  # Get the first listener that includes the given resource from worker
+  firstListenerOfType = type: ls: lib.lists.findFirst (isListenerType type)
+    (throw "No listener with resource: ${type} configured")
+    ls;
+
+  mainReplicationListener = firstListenerOfType "replication" cfg.settings.listeners;
+in {
+  options.services.matrix-synapse.workers = let
+    workerInstanceType = types.submodule ({ config, ... }: {
+      options.settings = mkOption {
+        type = types.submodule {
+          freeformType = format.type;
+
+          options = {
+            worker_app = mkOption {
+              type = types.enum [
+                "synapse.app.generic_worker"
+                "synapse.app.appservice"
+                "synapse.app.media_repository"
+                "synapse.app.user_dir"
```

Comment:
Only `generic_worker` and `media_repository` are not deprecated (see https://matrix-org.github.io/synapse/latest/workers.html#available-worker-applications), so `appservice` and `user_dir` should probably be removed.

Reply:
Thanks! There's a little bit of cleaning up to do for sure.